src/data/finset/noncomm_prod.lean

```diff
@@ -0,0 +1,167 @@
+  simp
+end
+
+/-- Fold of a `s : multiset α` with `op : α → α → α`, given a proof that `left_commutative op`
+on all elements `x ∈ s`. -/
+def noncomm_fold (s : multiset α) (comm) (a : α) : α :=
+noncomm_foldr op s comm a
+-- s.attach.foldr (f ∘ subtype.val) comm a
+
+@[simp] lemma noncomm_fold_coe (l : list α) (comm) (a : α) :
+  noncomm_fold op (l : multiset α) comm a = l.foldr op a :=
+by simp [noncomm_fold]
+
+@[simp] lemma noncomm_fold_empty (h) (a : α) :
+  noncomm_fold op (0 : multiset α) h a = a := rfl
+
+lemma noncomm_fold_cons (s : multiset α) (a : α) (h) (h') (x : α) :
+  noncomm_fold op (a ::ₘ s) h x = op a (noncomm_fold op s h' x) :=
+begin
+  induction s using quotient.induction_on,
+  simp
+end
+
+variables [monoid α]
+
+/-- Product of a `s : multiset α` with `[monoid α]`, given a proof that `*` commutes
+on all elements `x ∈ s`. -/
+@[to_additive "Sum of a `s : multiset α` with `[add_monoid α]`, given a proof that `*` commutes
+on all elements `x ∈ s`." ]
+def noncomm_prod (s : multiset α) (comm : ∀ (x ∈ s) (y ∈ s), commute x y) : α :=
+s.noncomm_fold (*) (λ x hx y hy b', by simp [←mul_assoc, (comm x hx y hy).eq]) 1
+
+@[simp, to_additive] lemma noncomm_prod_coe (l : list α) (comm) :
+  noncomm_prod (l : multiset α) comm = l.prod :=
+begin
+  rw [noncomm_prod],
+  simp only [noncomm_fold_coe],
+  induction l with hd tl hl,
+  { simp },
+  { rw [list.prod_cons, list.foldr, hl],
+    intros x hx y hy,
+    exact comm x (list.mem_cons_of_mem _ hx) y (list.mem_cons_of_mem _ hy) }
+end
+
+@[simp, to_additive] lemma noncomm_prod_empty (h) :
+  noncomm_prod (0 : multiset α) h = 1 := rfl
+
+@[simp, to_additive] lemma noncomm_prod_cons (s : multiset α) (a : α) (comm) :
+  noncomm_prod (a ::ₘ s) comm = a * noncomm_prod s
+    (λ x hx y hy, comm _ (mem_cons_of_mem hx) _ (mem_cons_of_mem hy)) :=
+begin
+  induction s using quotient.induction_on,
+  simp
+end
+
+@[to_additive] lemma noncomm_prod_cons' (s : multiset α) (a : α) (comm) :
+  noncomm_prod (a ::ₘ s) comm = noncomm_prod s
+    (λ x hx y hy, comm _ (mem_cons_of_mem hx) _ (mem_cons_of_mem hy)) * a :=
+begin
+  induction s using quotient.induction_on with s,
+  simp only [quot_mk_to_coe, cons_coe, noncomm_prod_coe, list.prod_cons],
+  induction s with hd tl IH,
+  { simp },
+  { rw [list.prod_cons, mul_assoc, ←IH, ←mul_assoc, ←mul_assoc],
+    { congr' 1,
+      apply comm;
+      simp },
+    { intros x hx y hy,
+      simp only [quot_mk_to_coe, list.mem_cons_iff, mem_coe, cons_coe] at hx hy,
+      apply comm,
+      { cases hx;
+        simp [hx] },
+      { cases hy;
+        simp [hy] } } }
+end
+
+end multiset
+
+namespace finset
+
+variables [monoid α]
+
+/-- Product of a `s : finset α` with `[monoid α]`, given a proof that `*` commutes
+on all elements `x ∈ s`. -/
+@[to_additive "Sum of a `s : finset α` with `[add_monoid α]`, given a proof that `*` commutes
+on all elements `x ∈ s`."]
+def noncomm_prod (s : finset α) (comm : ∀ (x ∈ s) (y ∈ s), commute x y) : α :=
```

Comment:
Probably this would be more useful in the style of `finset.prod`, where you have a function `f` that maps elements from `s` to elements in a monoid, and you're assuming `(comm : ∀ (x ∈ s) (y ∈ s), commute (f x) (f y))`